src/Storage/Record.php

```diff
@@ -0,0 +1,174 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Drupal\entity_hierarchy\Storage;
+
+use Drupal\Core\Entity\ContentEntityInterface;
+
+/**
+ * Defines a database record for a hierarchical query result.
+ */
+class Record {
+
+  protected int $id;
+
+  protected ?int $revisionId = NULL;
+
+  protected ?int $targetId;
+
+  protected int $weight;
+
+  protected int $depth;
+
+  protected ?array $children = NULL;
+
+  /**
+   * Constructor for a new Record.
+   *
+   * @param string $type
+   *   The entity type for this record.
+   */
+  public function __construct(
```

Comment:
I think we can make this private if we want people to use the ::create method

Reply:
Made it `protected` as part of a world domination plan :)

Reply:
PDO doesn't use the `::create` method. And the `__construct` is what allows direct instantiation with `$type` passed through. I think even if the tests pass with it protected, we should leave it public, because something is publicly calling it.